content/docs/patterns/template-method-pattern/index.md

```diff
@@ -0,0 +1,157 @@
+
+## Bad Code Example
+```al
+codeunit 50010 ExportSalesLines
+{
+    procedure ExportData(SalesHeader: Record "Sales Header", SalesLine : Record "Sales Line")
+    begin
+        if not SalesHeader.CheckData() then
+          exit;
+        repeat
+          case SalesHeader.ExportType of
+            Enum::ExportType::A:
+              GenerateLineTypeA(SalesLine);
+            Enum::ExportType::B:
+              GenerateLineTypeB(SalesLine);
+          end;
+        until SalesLine.Next() = 0;
+        
+        case SalesHeader.ExportType of
+          Enum::ExportType::A:
+            WriteToFile();
+          Enum::ExportType::B:
+            SendToWebService();
+        end;
+    end;
+
+    //TODO Implementation of procedures used in example
+}
+```
+As you can see in this example the readability gets bader with every new case.
+
+## The Pattern
+To implement the Pattern you need at least 3 objects:
+- A template codeunit
+- An Interface which provides the needed procedures
+- A codeunit which Implements the interrface
+
+In my Example I show how to implement a data export with templating
+
+We start with the template
+```al
+codeunit 50000 ExportTemplate
+{
+    procedure ExportData(export: Interface IDataExport)
+    begin
+        if not export.CheckData() then
+            exit;
+        if export.GetLinesToExport() then begin
+            repeat
+                export.ExportLine();
+            until not export.NextLine();
+        end;
+        export.Finish();
+    end;
+}
+```
+As you can see the template just calls procedures via an interface and just defines the flow of the export without really implementing it.
+
+As the second part we need an interface for the export functions
+```al
+interface IDataExport
+{
+    procedure CheckData(): Boolean;
+    procedure GetLinesToExport(): Boolean;
+    procedure ExportLine();
+    procedure NextLine(): Boolean;
+    procedure Finish();
+}
+```
+
+And now we need an implementation. For my example I wrote a export Codeunit for Sales Headers
+```al
+codeunit 50001 SalesHeaderExport implements IDataExport
+{
+    procedure SetSalesHeader(DocType: Enum "Sales Document Type"; No: Code[10])
+    begin
+        SalesHeader.Get(DocType, No);
+    end;
+
+    procedure CheckData(): Boolean
+    begin
+        SalesHeader.TestField(Status, Enum::"Sales Document Status"::Released);
+    end;
+
+    procedure GetLinesToExport(): Boolean
+    begin
+        SalesLines.SetRange("Document Type", SalesHeader."Document Type");
+        SalesLines.SetRange("Document No.", SalesHeader."No.");
+        exit(SalesLines.FindSet());
+    end;
+
+    procedure ExportLine()
+    begin
+        //Generate Exportdata here
+    end;
+
+    procedure NextLine(): Boolean
+    begin
+        exit(SalesLines.Next() > 0);
+    end;
```

Comment:
Should the template have a Direction (steps)? - possible as an overloaded procedure

```suggestion
    procedure NextLine(Steps : integer): Boolean
    begin
        exit(SalesLines.Next(Steps) <> 0);
    end;
```

Reply:
@JeremyVyska - You are marking all of my comments as resolved without any comment, isn't answering any questions. Please advise.